src/TorchSharp/DataLoader.cs

```diff
@@ -111,17 +121,50 @@ public bool MoveNext()
                         {
                             DisposeCurrent();
                             if (!MoveNextValue()) return false;
-                            List<Dictionary<string, Tensor>> dic = new();
-                            dic.Add(dataset.GetTensor(currentVal));
-                            for (var i = 1; i < batchSize; i++) {
+
+                            var tensorIndexList = new List<long> {currentVal};
+                            for (int i = 1; i < batchSize; i++) {
                                 if (!MoveNextValue()) break;
-                                dic.Add(dataset.GetTensor(currentVal));
+                                tensorIndexList.Add(currentVal);
                             }
 
+                            if (!shuffle)
+                                tensorIndexList.Reverse(); // ConcurrentBag::Add inserts data into first of the collections. So, I'll reverse them.
+
+                            ConcurrentBag<Dictionary<string, Tensor>> dic = new();
+                            var taskedBatchCount = 0;
+                            var taskBatchLock = new object();
+
+                            //Run Async
+                            foreach(var _ in Enumerable.Range(1, num_worker - 1))
+                                ThreadPool.QueueUserWorkItem(CreateBatch);
+
+                            CreateBatch(null);
+
+                            while (dic.Count < tensorIndexList.Count) { } //Wait for tensors ready
```

Comment:
While this is certainly a way to wait, you should be using a [`Semaphore`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.semaphore?view=net-6.0) instead.

You would not need a concurrent bag here then, an array would do.